wicket/src/rack_setup/config_toml.rs

```diff
@@ -0,0 +1,296 @@
+    array.set_trailing_comma(true);
+    array.set_trailing("\n");
+}
+
+fn build_sleds_array(sleds: &[BootstrapSledDescription]) -> Array {
+    let mut array = Array::new();
+    let mut prev: Option<&BootstrapSledDescription> = None;
+
+    for sled in sleds {
+        // We should never get a non-sled from wicketd; if we do, filter it out.
+        if sled.id.type_ != SpType::Sled {
+            continue;
+        }
+
+        let mut value = Formatted::new(i64::from(sled.id.slot));
+        let decor = value.decor_mut();
+
+        // We have to attach the comment for each sled on the _next_ item in the
+        // array, so here we set our prefix to be the previous item's details.
+        if let Some(prev) = prev {
+            decor.set_prefix(format!(
+                " # {serial} (model {model}, revision {rev}){sep}",
+                serial = prev.serial_number,
+                model = prev.model,
+                rev = prev.revision,
+                sep = ARRAY_SEP,
+            ));
+        } else {
+            decor.set_prefix(ARRAY_SEP);
+        }
+
+        array.push_formatted(Value::Integer(value));
+        prev = Some(sled);
+    }
+
+    // Because we attach comments to previous items, we also need to add a
+    // comment to the last element.
+    if let Some(prev) = prev {
+        array.set_trailing(format!(
+            " # {serial} (model {model}, revision {rev})\n",
+            serial = prev.serial_number,
+            model = prev.model,
+            rev = prev.revision,
+        ));
+        array.set_trailing_comma(true);
+    }
+
+    array
+}
+
+fn populate_network_table(
+    table: &mut Table,
+    config: Option<&RackNetworkConfig>,
+) {
+    // Helper function to serialize enums into their appropriate string
+    // representations.
+    fn enum_to_toml_string<T: Serialize>(value: &T) -> Cow<'static, str> {
+        let value = toml::Value::try_from(value).unwrap();
+        match value {
+            toml::Value::String(s) => Cow::from(s),
+            other => {
+                panic!("improper use of enum_to_toml_string: got {other:?}");
+            }
+        }
+    }
+
+    let Some(config) = config else {
+        return;
+    };
+
+    for (property, value) in [
+        ("gateway_ip", Cow::from(&config.gateway_ip)),
+        ("infra_ip_first", Cow::from(&config.infra_ip_first)),
+        ("infra_ip_last", Cow::from(&config.infra_ip_last)),
+        ("uplink_port", Cow::from(&config.uplink_port)),
+        ("uplink_port_speed", enum_to_toml_string(&config.uplink_port_speed)),
+        ("uplink_port_fec", enum_to_toml_string(&config.uplink_port_fec)),
+        ("uplink_ip", Cow::from(&config.uplink_ip)),
+    ] {
+        *table.get_mut(property).unwrap().as_value_mut().unwrap() =
+            Value::String(Formatted::new(value.into_owned()));
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use omicron_common::api::internal::shared::RackNetworkConfig as InternalRackNetworkConfig;
```

Comment:
Can this and `PortSpeed`/`InternalPortSpeed` share a type via progenitor's `replace`? Seems like a lot of extra ceremony converting between the types that isn't always necessary.

Reply:
Maybe! I'm still uneasy about when `replace` is dangerous or not, so didn't reach for it here...